File: packages/dsa-py/src/structures/lru_cache.py
"""
LRU Cache — O(1) en get, set y delete.
DoublyLinkedList + HashMap.
"""
from __future__ import annotations
from dataclasses import dataclass, field
from typing import Generic, TypeVar

K = TypeVar("K")
V = TypeVar("V")


@dataclass
class _Node(Generic[V]):
    key: str
    value: V
    prev: "_Node[V] | None" = field(default=None, repr=False)
    next: "_Node[V] | None" = field(default=None, repr=False)


class LRUCache(Generic[V]):
    """
    Args:
        capacity: número máximo de entradas
    """

    def __init__(self, capacity: int) -> None:
        if capacity <= 0:
            raise ValueError("capacity debe ser > 0")
        self._capacity = capacity
        self._map: dict[str, _Node[V]] = {}
        # sentinel head y tail para simplificar la lógica
        self._head: _Node[V] = _Node(key="__head__", value=None)  # type: ignore[arg-type]
        self._tail: _Node[V] = _Node(key="__tail__", value=None)  # type: ignore[arg-type]
        self._head.next = self._tail
        self._tail.prev = self._head
        self._hits   = 0
        self._misses = 0

    @property
    def size(self) -> int:
        return len(self._map)

    def get(self, key: str) -> V | None:
        """O(1)"""
        node = self._map.get(key)
        if node is None:
            self._misses += 1
            return None
        self._move_to_front(node)
        self._hits += 1
        return node.value

    def set(self, key: str, value: V) -> None:
        """O(1)"""
        if key in self._map:
            node = self._map[key]
            node.value = value
            self._move_to_front(node)
        else:
            if len(self._map) >= self._capacity:
                self._evict_lru()
            node = _Node(key=key, value=value)
            self._map[key] = node
            self._insert_front(node)

    def delete(self, key: str) -> bool:
        """O(1)"""
        node = self._map.pop(key, None)
        if node is None:
            return False
        self._remove_node(node)
        return True

    def stats(self) -> dict[str, int | float]:
        total = self._hits + self._misses
        return {
            "hits":    self._hits,
            "misses":  self._misses,
            "size":    self.size,
            "hit_rate": round(self._hits / total * 100) if total else 0,
        }

    def _insert_front(self, node: _Node[V]) -> None:
        node.prev = self._head
        node.next = self._head.next
        if self._head.next:
            self._head.next.prev = node
        self._head.next = node

    def _remove_node(self, node: _Node[V]) -> None:
        if node.prev:
            node.prev.next = node.next
        if node.next:
            node.next.prev = node.prev

    def _move_to_front(self, node: _Node[V]) -> None:
        self._remove_node(node)
        self._insert_front(node)

    def _evict_lru(self) -> None:
        lru = self._tail.prev
        if lru and lru is not self._head:
            self._remove_node(lru)
            del self._map[lru.key]

File: packages/dsa-ts/src/structures/LRUCache.ts
import { DoublyLinkedList, ListNode } from './LinkedList.js';

interface CacheEntry<V> {
  key:       string;
  value:     V;
  expiresAt: number | null;   // null = sin TTL
}

interface CacheStats {
  hits:    number;
  misses:  number;
  size:    number;
  hitRate: number;
}

/**
 * LRU Cache (Least Recently Used)
 *
 * Combina DoublyLinkedList + HashMap para lograr O(1) en todas las ops.
 * El nodo más recientemente usado va al frente de la lista.
 * Cuando se llena, se evicta el nodo al final (menos usado).
 *
 * Complexity:
 *   get    O(1)
 *   set    O(1)
 *   delete O(1)
 */
export class LRUCache<V> {
  private readonly list = new DoublyLinkedList<CacheEntry<V>>();
  private readonly map  = new Map<string, ListNode<CacheEntry<V>>>();
  private hits   = 0;
  private misses = 0;

  constructor(private readonly capacity: number) {
    if (capacity <= 0) throw new RangeError('capacity debe ser > 0');
  }

  /**
   * Recupera un valor. Mueve el nodo al frente (más reciente). O(1)
   * @returns el valor o undefined si no existe / expiró
   */
  get(key: string): V | undefined {
    const node = this.map.get(key);
    if (node === undefined) {
      this.misses++;
      return undefined;
    }

    const entry = node.value;

    // comprobar TTL
    if (entry.expiresAt !== null && Date.now() > entry.expiresAt) {
      this.evict(key, node);
      this.misses++;
      return undefined;
    }

    // mover al frente (más reciente)
    this.list.remove(node);
    const fresh = this.list.prepend(entry);
    this.map.set(key, fresh);
    this.hits++;
    return entry.value;
  }

  /**
   * Inserta o actualiza un valor. Evicta el LRU si está lleno. O(1)
   * @param ttlMs milisegundos de vida — undefined = sin expiración
   */
  set(key: string, value: V, ttlMs?: number): void {
    const existing = this.map.get(key);
    if (existing !== undefined) {
      this.list.remove(existing);
      this.map.delete(key);
    } else if (this.list.size >= this.capacity) {
      // evictar el LRU (tail = menos usado)
      const lruNode = this.list.tail;
      if (lruNode !== null) {
        this.evict(lruNode.value.key, lruNode);
      }
    }

    const entry: CacheEntry<V> = {
      key,
      value,
      expiresAt: ttlMs !== undefined ? Date.now() + ttlMs : null,
    };
    const node = this.list.prepend(entry);
    this.map.set(key, node);
  }

  /** Elimina una clave explícitamente. O(1) */
  delete(key: string): boolean {
    const node = this.map.get(key);
    if (node === undefined) return false;
    this.evict(key, node);
    return true;
  }

  has(key: string): boolean {
    return this.map.has(key) && this.get(key) !== undefined;
  }

  get size(): number { return this.list.size; }

  /** Estadísticas de uso */
  stats(): CacheStats {
    const total = this.hits + this.misses;
    return {
      hits:    this.hits,
      misses:  this.misses,
      size:    this.size,
      hitRate: total === 0 ? 0 : Math.round((this.hits / total) * 100),
    };
  }

  clear(): void {
    this.list.head = null;
    this.list.tail = null;
    this.map.clear();
    this.hits   = 0;
    this.misses = 0;
  }

  private evict(key: string, node: ListNode<CacheEntry<V>>): void {
    this.list.remove(node);
    this.map.delete(key);
  }
}
